File: common_book/src/txtmfqbyd/java/com/intelligent/reader/widget/drawer/DrawerLayout.java
package com.intelligent.reader.widget.drawer;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.v4.os.ParcelableCompat;
import android.support.v4.os.ParcelableCompatCreatorCallbacks;
import android.support.v4.view.AbsSavedState;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.ViewDragHelper;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;

import swipeback.ShadowView;

/**
 * Desc 抽屉菜单布局
 * Author qiantao
 * Mail tao_qian@dingyuegroup.cn
 * Date 2018/2/24
 */

public class DrawerLayout extends FrameLayout {

    private static final String TAG = "SRDrawerLayout";
    private final int screenWidth;

    private View menuView;
    private DrawerMain mainView;
    private final ShadowView shadowView;

    private final ViewDragHelper viewDragHelper;

    private static final int MENU_CLOSED = 1;
    private static final int MENU_OPENED = 2;
    private int menuState = MENU_CLOSED;

    private int dragOrientation;
    private static final int LEFT_TO_RIGHT = 3;
    private static final int RIGHT_TO_LEFT = 4;

    private static final float SPRING_BACK_VELOCITY = 1500;
    private static final int SPRING_BACK_DISTANCE = 80;
    private final int springBackDistance;

    private static final int MENU_MARGIN_RIGHT = 64;
    private final int menuWidth;

    private static final int SHADOW_WIDTH = 15;
    private final int shadowWidth;

    private static final float TOUCH_SLOP_SENSITIVITY = 1.f;

    private int mainLeft;

    public DrawerLayout(Context context) {
        this(context, null);
    }

    public DrawerLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public DrawerLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        final float density = getResources().getDisplayMetrics().density;//屏幕密度

        screenWidth = getResources().getDisplayMetrics().widthPixels;

        springBackDistance = (int) (SPRING_BACK_DISTANCE * density + 0.5f);

        mainLeft = 0;

        menuWidth = screenWidth - (int) (MENU_MARGIN_RIGHT * density + 0.5f);

        shadowView = new ShadowView(context);
        shadowWidth = (int) (SHADOW_WIDTH * density + 0.5f);

        viewDragHelper = ViewDragHelper.create(this, TOUCH_SLOP_SENSITIVITY, new CoordinatorCallback());
        viewDragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_LEFT);
    }

    private class CoordinatorCallback extends ViewDragHelper.Callback {

        @Override
        public void onEdgeDragStarted(int edgeFlags, int pointerId) {
            viewDragHelper.captureChildView(mainView, pointerId);
        }

        @Override
        public boolean tryCaptureView(View child, int pointerId) {
            if (menuState == MENU_OPENED) {
                return menuView == child || mainView == child;
            } else {
                return menuView == child;
            }
        }

        @Override
        public void onViewCaptured(View capturedChild, int activePointerId) {
            if (capturedChild == menuView) {
                viewDragHelper.captureChildView(mainView, activePointerId);
            }
        }

        @Override
        public int getViewHorizontalDragRange(View child) {
            return 1;
        }

        @Override
        public int clampViewPositionHorizontal(View child, int left, int dx) {
            if (left < 0) {
                left = 0;
            } else if (left > menuWidth) {
                left = menuWidth;
            }
            return left;
        }

        @Override
        public void onViewReleased(View releasedChild, float xvel, float yvel) {
            super.onViewReleased(releasedChild, xvel, yvel);
//            Log.e(TAG, "onViewReleased: xvel: " + xvel);
            if (dragOrientation == LEFT_TO_RIGHT) {
                if (xvel > SPRING_BACK_VELOCITY || mainView.getLeft() > springBackDistance) {
                    openMenu();
                } else {
                    closeMenu();
                }
            } else if (dragOrientation == RIGHT_TO_LEFT) {
                if (xvel < -SPRING_BACK_VELOCITY || mainView.getLeft() < menuWidth - springBackDistance) {
                    closeMenu();
                } else {
                    openMenu();
                }
            }

        }

        @Override
        public void onViewPositionChanged(View changedView, int left, int top, int dx, int dy) {
//            Log.d(TAG, "onViewPositionChanged: left:" + left);
            mainLeft = left;
            if (dx > 0) {
                dragOrientation = LEFT_TO_RIGHT;
            } else if (dx < 0) {
                dragOrientation = RIGHT_TO_LEFT;
            }
        }
    }

    //加载完布局文件后调用
    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        menuView = getChildAt(0);

        mainView = (DrawerMain) getChildAt(1);
        mainView.setParent(this);

        addView(shadowView);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {
        return viewDragHelper.shouldInterceptTouchEvent(event);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        //将触摸事件传递给ViewDragHelper，此操作必不可少
        viewDragHelper.processTouchEvent(event);
        return true;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);

        MarginLayoutParams menuParams = (MarginLayoutParams) menuView.getLayoutParams();
        menuParams.width = menuWidth;
        menuView.setLayoutParams(menuParams);

        mainView.layout(mainLeft, 0, mainLeft + screenWidth, bottom);

        MarginLayoutParams shadowParams = (MarginLayoutParams) shadowView.getLayoutParams();
        shadowParams.width = shadowWidth;
        shadowView.setLayoutParams(shadowParams);
        int shadowLeft = mainLeft - shadowWidth;
        shadowView.layout(shadowLeft, 0, mainLeft, bottom);
    }

    @Override
    protected boolean drawChild(Canvas canvas, View child, long drawingTime) {
        final int restoreCount = canvas.save();//保存画布当前的剪裁信息

        int clipLeft = 0;
        int clipRight = mainView.getLeft();
        if (child == menuView && clipRight == 0) {//在 menu 关闭后剪裁
            canvas.clipRect(clipLeft, 0, clipRight, getHeight());//剪裁显示的区域
        }

        boolean result = super.drawChild(canvas, child, drawingTime);//绘制当前view

        //恢复画布之前保存的剪裁信息
        //以正常绘制之后的view
        canvas.restoreToCount(restoreCount);

        shadowView.setX(mainLeft - shadowWidth);

        return result;
    }

    @Override
    public void computeScroll() {
        if (viewDragHelper.continueSettling(true)) {
            ViewCompat.postInvalidateOnAnimation(this);
        }
        if (mainView.getLeft() == 0) {
            menuState = MENU_CLOSED;
        } else if (mainView.getLeft() == menuWidth) {
            menuState = MENU_OPENED;
        }
    }

    public void openMenu() {
        viewDragHelper.smoothSlideViewTo(mainView, menuWidth, 0);
        ViewCompat.postInvalidateOnAnimation(DrawerLayout.this);
    }

    public void closeMenu() {
        viewDragHelper.smoothSlideViewTo(mainView, 0, 0);
        ViewCompat.postInvalidateOnAnimation(DrawerLayout.this);
    }

    public boolean isOpened() {
        return menuState == MENU_OPENED;
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        final Parcelable superState = super.onSaveInstanceState();
        final SavedState ss = new SavedState(superState);
        ss.menuState = menuState;
        return ss;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (!(state instanceof SavedState)) {
            super.onRestoreInstanceState(state);
            return;
        }

        final SavedState ss = (SavedState) state;
        super.onRestoreInstanceState(ss.getSuperState());

        if (ss.menuState == MENU_OPENED) {
            openMenu();
        }
    }

    protected static class SavedState extends AbsSavedState {
        int menuState;

        SavedState(Parcel in, ClassLoader loader) {
            super(in, loader);
            menuState = in.readInt();
        }

        SavedState(Parcelable superState) {
            super(superState);
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            super.writeToParcel(dest, flags);
            dest.writeInt(menuState);
        }

        public static final Creator<SavedState> CREATOR = ParcelableCompat.newCreator(
                new ParcelableCompatCreatorCallbacks<SavedState>() {
                    @Override
                    public SavedState createFromParcel(Parcel in, ClassLoader loader) {
                        return new SavedState(in, loader);
                    }

                    @Override
                    public SavedState[] newArray(int size) {
                        return new SavedState[size];
                    }
                });
    }

    public int transToHsvColor(int color) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        hsv[1] = hsv[1] + 0.1f;
        hsv[2] = hsv[2] - 0.1f;
        return Color.HSVToColor(hsv);
    }
}

File: common_book/src/txtqbmfyd/java/com/intelligent/reader/widget/drawer/DrawerLayout.kt
package com.intelligent.reader.widget.drawer

import android.content.Context
import android.graphics.Canvas
import android.os.Parcel
import android.os.Parcelable
import android.support.v4.os.ParcelableCompat
import android.support.v4.os.ParcelableCompatCreatorCallbacks
import android.support.v4.view.AbsSavedState
import android.support.v4.view.ViewCompat
import android.support.v4.widget.ViewDragHelper
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.intelligent.reader.widget.drawer.DrawerLayout.MenuState.MENU_CLOSED
import com.intelligent.reader.widget.drawer.DrawerLayout.MenuState.MENU_OPENED
import net.lzbook.kit.utils.AppLog


/**
 * Desc 抽屉菜单布局
 * Author qiantao
 * Mail tao_qian@dingyuegroup.cn
 * Date 2018/2/24
 */

class DrawerLayout @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0)
    : FrameLayout(context, attrs, defStyleAttr) {
    private val screenWidth: Int
    private val screenHeight: Int

    private var menuView: View? = null
    private var mainView: DrawerMain? = null
    private val shadowView: ShadowView

    private val viewDragHelper: ViewDragHelper
    private var menuState: Int = MENU_CLOSED

    private var dragOrientation: Int = 0
    private val springBackDistance: Int
    private val menuWidth: Int
    private val shadowWidth: Int

    private var mainLeft: Int = 0

    private var isLock = false

    val isOpened: Boolean
        get() = menuState == MENU_OPENED

    private var onMenuStateChangeListener: ((menuState: Int) -> Unit)? = null

    init {

        val density = resources.displayMetrics.density//屏幕密度

        screenWidth = resources.displayMetrics.widthPixels
        screenHeight = resources.displayMetrics.heightPixels

        springBackDistance = (SPRING_BACK_DISTANCE * density + 0.5f).toInt()

        mainLeft = 0

        menuWidth = screenWidth - (MENU_MARGIN_RIGHT * density + 0.5f).toInt()

        shadowView = ShadowView(context)
        shadowWidth = (SHADOW_WIDTH * density + 0.5f).toInt()

        viewDragHelper = ViewDragHelper.create(this, TOUCH_SLOP_SENSITIVITY, CoordinatorCallback())
        viewDragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_LEFT)
    }

    private inner class CoordinatorCallback : ViewDragHelper.Callback() {

        override fun onEdgeDragStarted(edgeFlags: Int, pointerId: Int) {
            viewDragHelper.captureChildView(mainView, pointerId)
        }

        override fun tryCaptureView(child: View, pointerId: Int): Boolean {
            return if (menuState == MENU_OPENED) {
                menuView === child || mainView === child
            } else {
                menuView === child
            }
        }

        override fun onViewCaptured(capturedChild: View?, activePointerId: Int) {
            if (capturedChild === menuView) {
                viewDragHelper.captureChildView(mainView, activePointerId)
            }
        }

        override fun getViewHorizontalDragRange(child: View): Int {
            return measuredWidth -child.measuredWidth
        }

        override fun clampViewPositionHorizontal(child: View?, left: Int, dx: Int): Int {
            var l = left
            if (l < 0) {
                l = 0
            } else if (l > menuWidth) {
                l = menuWidth
            }
            return if (isLock) {
                0
            } else {
                l
            }
        }

        override fun onViewReleased(releasedChild: View?, xvel: Float, yvel: Float) {
            super.onViewReleased(releasedChild, xvel, yvel)
            //            Log.e(TAG, "onViewReleased: xvel: " + xvel);
            if (dragOrientation == LEFT_TO_RIGHT) {
                if (xvel > SPRING_BACK_VELOCITY || mainView?.left ?: 0 > springBackDistance) {
                    openMenu()
                } else {
                    closeMenu()
                }
            } else if (dragOrientation == RIGHT_TO_LEFT) {
                if (xvel < -SPRING_BACK_VELOCITY || mainView?.left ?: 0 < menuWidth - springBackDistance) {
                    closeMenu()
                } else {
                    openMenu()
                }
            }

        }

        override fun onViewPositionChanged(changedView: View?, left: Int, top: Int, dx: Int, dy: Int) {
            mainLeft = left

            moveView(screenHeight)

            if (dx > 0) {
                dragOrientation = LEFT_TO_RIGHT
            } else if (dx < 0) {
                dragOrientation = RIGHT_TO_LEFT
            }
        }
    }

    //加载完布局文件后调用
    override fun onFinishInflate() {
        super.onFinishInflate()
        menuView = getChildAt(0)
        val menuParams = menuView?.layoutParams as ViewGroup.MarginLayoutParams
        menuParams.width = menuWidth
        menuView?.layoutParams = menuParams

        mainView = getChildAt(1) as DrawerMain
        mainView?.setParent(this)

        val shadowParams = FrameLayout.LayoutParams(shadowWidth, FrameLayout.LayoutParams.MATCH_PARENT)
        addView(shadowView, shadowParams)
    }

    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
        AppLog.e(TAG, "onInterceptTouchEvent")
        return viewDragHelper.shouldInterceptTouchEvent(event)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        //将触摸事件传递给ViewDragHelper，此操作必不可少
        AppLog.e(TAG, "onTouchEvent")
        viewDragHelper.processTouchEvent(event)
        return true
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)

        moveView(bottom)
    }

    override fun drawChild(canvas: Canvas, child: View, drawingTime: Long): Boolean {
        val restoreCount = canvas.save()//保存画布当前的剪裁信息

        val clipLeft = 0
        val clipRight = mainView?.left ?: 0
        if (child === menuView && clipRight == 0 && menuState == MENU_CLOSED) {
            canvas.clipRect(clipLeft, 0, clipRight, height)//剪裁显示的区域
        }

        val result = super.drawChild(canvas, child, drawingTime)//绘制当前view

        //恢复画布之前保存的剪裁信息
        //以正常绘制之后的view
        canvas.restoreToCount(restoreCount)

        return result
    }

    override fun computeScroll() {
        if (viewDragHelper.continueSettling(true)) {
            ViewCompat.postInvalidateOnAnimation(this)
        }
        if (mainView?.left == 0 && menuState != MENU_CLOSED) {
            menuState = MENU_CLOSED
            onMenuStateChangeListener?.invoke(MENU_CLOSED)
        } else if (mainView?.left == menuWidth && menuState != MENU_OPENED) {
            menuState = MENU_OPENED
            onMenuStateChangeListener?.invoke(MENU_OPENED)
        }
    }

    fun openMenu() {
        viewDragHelper.smoothSlideViewTo(mainView, menuWidth, 0)
        ViewCompat.postInvalidateOnAnimation(this@DrawerLayout)
    }

    fun closeMenu() {
        viewDragHelper.smoothSlideViewTo(mainView, 0, 0)
        ViewCompat.postInvalidateOnAnimation(this@DrawerLayout)
    }

    fun setOnMenuStateChangeListener(listener: (state: Int) -> Unit) {
        onMenuStateChangeListener = listener
    }

    fun lock() {
        isLock = true
    }

    fun unlock() {
        isLock = false
    }

    override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        val ss = SavedState(superState)
        ss.menuState = menuState
        return ss
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }

        super.onRestoreInstanceState(state.superState)

        if (state.menuState == MENU_OPENED) {
            openMenu()
        }
    }

    protected class SavedState : AbsSavedState {
        internal var menuState: Int = 0

        internal constructor(`in`: Parcel, loader: ClassLoader) : super(`in`, loader) {
            menuState = `in`.readInt()
        }

        internal constructor(superState: Parcelable) : super(superState) {}

        override fun writeToParcel(dest: Parcel, flags: Int) {
            super.writeToParcel(dest, flags)
            dest.writeInt(menuState)
        }

        val CREATOR: Parcelable.Creator<SavedState> = ParcelableCompat.newCreator(
                object : ParcelableCompatCreatorCallbacks<SavedState> {
                    override fun createFromParcel(`in`: Parcel, loader: ClassLoader): SavedState {
                        return SavedState(`in`, loader)
                    }

                    override fun newArray(size: Int): Array<SavedState?> {
                        return arrayOfNulls(size)
                    }
                })

    }

    private fun moveView(bottom: Int) {
        mainView?.layout(mainLeft, 0, mainLeft + screenWidth, bottom)

        val shadowLeft = mainLeft - shadowWidth
        shadowView.layout(shadowLeft, 0, mainLeft, bottom)
    }

    companion object {

        private val TAG = "DrawerLayout"

        private val LEFT_TO_RIGHT = 3
        private val RIGHT_TO_LEFT = 4

        private val SPRING_BACK_VELOCITY = 1500f
        private val SPRING_BACK_DISTANCE = 80

        private val MENU_MARGIN_RIGHT = 100

        private val SHADOW_WIDTH = 20

        private val TOUCH_SLOP_SENSITIVITY = 1f
    }

    object MenuState {
        val MENU_CLOSED = 1
        val MENU_OPENED = 2
    }

}